python/tskit/combinatorics.py

```diff
@@ -0,0 +1,798 @@
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+Module for ranking and unranking trees. Trees are considered only
+leaf-labelled and unordered, so order of children does not influence equality.
+"""
+import heapq
+from functools import lru_cache
+
+import tskit
+
+
+def rank(tree):
+    """
+    The rank of a tree is composed of two parts:
+    1. The shape rank - its rank in the enumeration of all unlabelled
+        trees of n leaves.
+    2. The label rank - its rank in the enumeration of all labelings
+        of this tree.
+    """
+    return RankTree.from_tsk_tree(tree).rank()
+
+
+def shape_rank(tree):
+    return RankTree.from_tsk_tree(tree).shape_rank()
+
+
+def label_rank(tree):
+    return RankTree.from_tsk_tree(tree).label_rank()
+
+
+def unrank(rank, num_leaves):
+    """
+    Reconstruct the tree of the given ``rank`` (see :func: `tskit.combinatorics.rank`)
+    with ``num_leaves`` leaves. The labels and times of internal nodes are chosen
+    arbitrarily, and the time of each leaf is 0.
+
+    :param tuple(int) rank: The rank of the tree to generate.
+    :param int num_leaves: The number of leaves of the tree to generate.
+    :rtype: Tree
+    """
+    return RankTree.unrank(rank, num_leaves).to_tsk_tree()
+
+
+def shape_unrank(shape_rank, num_leaves):
+    """
+    Reconstruct the tree of the given shape rank ``shape_rank``
+    (see :func: `tskit.combinatorics.rank`) with ``num_leaves`` leaves.
+    Leaves are labelled in increasing order from left to right.
+
+    :param int shape_rank: The shape rank of the tree to generate.
+    :param int num_leaves: The number of leaves of the tree to generate.
+    :rtype: Tree
+    """
+    return RankTree.unrank((shape_rank, 0), num_leaves).to_tsk_tree()
+
+
+def label_unrank(tree, label_rank):
+    """
+    Produce a tree identical in shape to ``tree`` with the labelling
+    corresponding to the given label rank ``rank``
+    (see :func: `tskit.combinatorics.rank`).
+
+    :param Tree tree: The tree to relabel.
+    :param int label_rank: The rank of the labelling of ``tree`` to use.
+    :rtype: Tree
+    """
+    return RankTree.from_tsk_tree(tree).label_unrank(label_rank).to_tsk_tree()
+
+
+class RankTree:
+    """
+    A tree class that maintains the topological ranks of each node in the tree.
+    This structure can be used to efficiently compute the rank of a tree of
+    n leaves and produce a tree given a rank.
+    """
+
+    def __init__(self, shape_rank=None, label_rank=None, children=None, label=None):
```

Comment:
The boundary conditions don't feel quite right to me, and the behaviour seems quite complicated. How are you constructing RankTrees in practise? It's seems ugly that you have all these special cases for n = 1, when that should just fall out if the conditions have been set up properly. It would be good to make a single constructor with mandatory arguments.

Reply:
Ah yes the `n = 1` cases seemed to mostly be residue from my recursive implementations and were pretty easily resolved. The constructor is pretty messy and I found myself having two cases of constructing RankTrees that seemed at odds regarding the ranks (children is easy to make mandatory). When going from a tskit Tree, we need to compute the ranks as we construct the nodes from leaves to root. But when unranking, we already know the ranks of the children as part of the unranking process. One way to avoid this mess entirely could be to just make the ranks lazy, cached properties.

Reply:
The reason for doing it this way is that in order to label the trees you need the full tree of shape ranks, which I was trying to avoid recomputing.